smolyay/samples.py

```diff
@@ -0,0 +1,496 @@
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    num_level: int
+        The number of levels.
+    """
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating nested extrema of chebyshev polynomials of the first kind.
+        """
+        # create properties for levels, level 0 is a special case with 1 point
+        rule = lambda x: 1 if x == 0 else int(2 ** (numpy.ceil(numpy.log2(x)) + 1) + 1)
+        self._num_per_level = numpy.ones(self.num_level, dtype=int)
+        self._num_per_level[1:] = [
+            rule(i) - rule((i - 1)) for i in range(1, self.num_level)
+        ]
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+        # points, level 0 is a special case only 0 as a point
+        num_points = self._end_level[-1]
+        points = numpy.zeros(num_points, dtype=float)
+        for i in range(1, self.num_level):
+            if self._num_per_level[i] == 0:
+                continue
+            # find indexes of extrema not already found. Fraction index/degree
+            # cannot be further simplified
+            degree = int(2 ** (numpy.ceil(numpy.log2(i)) + 1))
+            if i == 1:
+                # special case for level == 1
+                indexes = numpy.arange(0, degree + 1, 2, dtype=int)
+            else:
+                indexes = indexes = numpy.arange(1, degree, 2, dtype=int)
+                indexes = indexes[~(numpy.gcd(indexes, degree) > 1)]
+            points[self._start_level[i] : self._end_level[i]] = -numpy.cos(
+                numpy.pi * indexes / degree
+            )
+        self._points = self._scale_to_domain(points, [-1, 1])
+
+
+class TrigonometricPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Trigonometric sampling
+
+    The :attr:`points` for this interpolation scheme are the
+    trigonometric interpolation points for m points
+
+    .. math::
+
+        x^f_j = \frac{2\pi j}{m(f)},  1 \leq j \leq m(l), f \geq 0
+
+    where f is the frequency. The relationship between the frequency
+    and the number of points is set at 
+    
+    .. math::
+        `m = 2*\left | f \right | + 1` 
+    
+    to ensure that m is an odd number at every frequency.
+
+    The points are then scaled from the domain :math:`[0, 2\pi]`
+    to the domain specified by the parameter `domain`.
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    frequency: int
+        The frequency to take points from.
+    """
+
+    def __init__(self, domain, frequency):
+        super().__init__(domain)
+        self._frequency = None
+
+        self.frequency = frequency
+
+    @property
+    def frequency(self):
+        """int: frequency to create points for."""
+        return self._frequency
+
+    @frequency.setter
+    def frequency(self, value):
+        frequency = int(value)
```

Comment:
Should we just document that `frequency` should be positive, and then either take the absolute value here to be sure or else error if it isn't? Then we don't have to do it below.

Reply:
It shouldn't error, since trigonometric polynomial frequencies are allowed to be negative. It can be forced to the absolute value.

Reply:
I am worried this behavior will be confusing because when the user enters a negative frequency, the object will give back a positive frequency.

For the purposes of generating a point set, I think it is a good idea to just require a positive frequency, especially now that we are generating the points in pairs. The negative frequency is more meaningful for the basis functions, and the user can always take the absolute value themselves if they want to work with a negative frequency. This is consistent with Python’s “explicit is better than implicit” concept.

Please require frequency to be nonnegative, and make this an error otherwise!

Reply:
Done